### PR TITLE
SNOW-281305 implement bulk array binding

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
[SNOW-281305](https://snowflakecomputing.atlassian.net/browse/SNOW-281305?atlOrigin=eyJpIjoiYzY4MTg1MTJkOTIwNDFjMGExZjQ4NTEzOGEyZTJmZDEiLCJwIjoiaiJ9)

### CODE REVIEW

File: src/snowflake/connector/cursor.py [hunk 1]
```diff
@@ -436,7 +440,7 @@ def interrupt_handler(*_):  # pragma: no cover
 
     def execute(self,
                 command: str,
-                params: Union[List, Tuple, None] = None,
+                params: Union[List, Tuple, None, str] = None,
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/cursor.py:443]: Split this into a private argument instead as per our conversation.
In that case will this work with `pyformat` too?

File: test/integ/test_bindings.py [hunk 2]
```diff
@@ -287,33 +306,40 @@ def test_binding_bulk_insert(conn_cnx, db_parameters):
...
 drop table if exists {name}
 """.format(name=db_parameters['name']))
 
 
+@pytest.mark.skipolddriver
+def test_bulk_insert_binding_fallback(conn_cnx, db_parameters):
+    """When stage creation fails, bulk inserts falls back to server side binding and disables stage optimization."""
+    with conn_cnx(paramstyle='qmark') as cnx, cnx.cursor() as csr:
+        query = 'insert into {name}(c1,c2) values(?,?)'.format(
+            name=db_parameters['name'])
```
Comment (sfc-gh-mkeller) [inline test/integ/test_bindings.py:328]: Use randomized name generator instead

File: test/unit/test_bind_upload_agent.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env python
+
```
Comment (sfc-gh-mkeller) [inline test/unit/test_bind_upload_agent.py:2]:
```suggestion
```

**Review by sfc-gh-mkeller — CHANGES_REQUESTED** [review]

File: src/snowflake/connector/bind_upload_agent.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        self.rows = rows
+        self._stream_buffer_size = stream_buffer_size
+        self.stage_path = f"@{_STAGE_NAME}/{uuid.uuid4().hex}"
+
+    def _create_stage(self):
+        self.cursor.execute(_CREATE_STAGE_STMT)
+
+    def upload(self):
+        try:
+            self._create_stage()
```
[the remark this answers is not in the document]
Comment (sfc-gh-stan) [reply src/snowflake/connector/bind_upload_agent.py:43]: I separated this into two functions for testing, where I patch the stage creation to fail but does not wrap the upload logic.

File: src/snowflake/connector/cursor.py (hunk 1 above)
Comment (sfc-gh-stan) [reply src/snowflake/connector/cursor.py]: I don't think so, `pyformat` always uses client side binding, I'm not sure if GS will parse the query correctly and do `COPY INTO` even if we upload bindings as files.

**Review by sfc-gh-mkeller — APPROVED** [review]
🚢 🚢

**Pull request merged**